src/main/java/net/dries007/tfc/common/recipes/ingredients/TraitIngredient.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the EUPL, Version 1.2.
+ * You may obtain a copy of the Licence at:
+ * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ */
+
+package net.dries007.tfc.common.recipes.ingredients;
+
+import java.util.function.BiFunction;
+
+import com.google.gson.JsonObject;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+
+import net.minecraftforge.common.crafting.IIngredientSerializer;
+
+import net.dries007.tfc.common.capabilities.food.FoodCapability;
+import net.dries007.tfc.common.capabilities.food.FoodTrait;
+import net.dries007.tfc.util.Helpers;
+import net.dries007.tfc.util.JsonHelpers;
+import org.jetbrains.annotations.Nullable;
+
+public abstract class TraitIngredient extends DelegateIngredient
+{
+    private final FoodTrait trait;
```

Comment:
Could make this `protected` rather than requiring subclasses to call `getTrait()`